doc/design.md

```diff
@@ -0,0 +1,1268 @@
+Therefore we need to ensure that the API is reasonably stable.
+At the same time, we will struggle with things that do not work out early on as imagined right now and will need to change the API.
+- The API is considered unstable and can change without notice until the library version has reached v1.0.
+- With v1.0 onwards the Python API should only change with major releases and a proper deprecation cycle.
+- The C++ API should ideally follow the same rules as the Python API.
+  However, since most or all C++ code using the API should be under our control, the API may be updated in a less restrictive way if there is a good reason.
+
+The implication is that this library would probably require versioning independent of Mantid itself.
+
+
+### <a name="design-details-naming"></a>Naming
+
+#### Components of the library are part of namespace `dataset`
+
+##### Rationale
+
+Mantid currently uses upper-case namespace names.
+However, this can lead to artificial and unintuitive namespace or class names.
+Instead, namespaces like `dataset` should be lower-case, as opposed to types which are upper-case.
+We then have, e.g., `dataset::Dataset`.
+This also matches the Python convention where module names are typically lower-case, thus reducing the differences between the C++ and Python APIs.
+
+#### The Python module name is `dataset`
+
+##### Rationale
+
+This is a corollary from the namespace name.
+
+#### The top level namespace and module is `mantid`
+
+##### Rationale
+
+Currently we have `namespace Mantid` in C++ while the Python module is `mantid`.
+Changing the namespace name would improve consistency.
+Furthermore, this would reduce naming clashes between existing and new code.
+
+##### Questions
+
+Does this make sense?
+Where to we put additional functionality?
+Sibling or children of `dataset`?
+Consider `class Dataset` itself and two hypothetical modules with *additional* functionality that is not part of this library.
+Our options would include:
+1. `mantid::dataset::Dataset`, `mantid::dataset::algorithm`, `mantid::dataset::io`.
+1. `mantid::dataset::Dataset`, `mantid::algorithm`, `mantid::io`.
+1. `mantid::dataset::core::Dataset`, `mantid::dataset::algorithm`, `mantid::dataset::io`.
+
+
+### <a name="design-details-attributes"></a>Attributes
+
+##### Context
+
+Variables grouped into coordinates and data cover almost all relevant cases, both in terms of what information needs to be stored as well as the behavior under operations, matching for coordinates and transformation for data.
+However, there is one exception:
+Metadata seems to fall in neither of the two categories.
+In the case of Mantid almost all metadata is contained in the `API::Run` object.
+The current behavior is to carry the metadata through most operations unchanged, with a few exceptions like `Algorithms::Plus` which merge the metadata.
```

Comment:
:+1: 
Are all the things you mention already supported in Mantid, or is there a requirement to do more, or more flexibly?

Reply:
It's mostly supported already. The only thing missing in `MergeRuns` is the possibility to set the `run_start` to the earliest time stamp and `run_end` to the last one found in the set of datasets-to-merge, but otherwise the merging rules are very flexible. See the [docs](http://docs.mantidproject.org/nightly/algorithms/MergeRuns-v1.html#merging-sample-logs) for more details.

Reply:
So this behavior needs to be configurable at run-time? That may make things more tricky. On the other hand, it is mostly encapsulated in a single algorithm, so I suppose we should be able to support all this also in the new design.

One thing I had in mind:

`API::Run` is actually very similar to `Dataset`, i.e., instead of storing `API::Run` in a variable, we store a (nested) `Dataset`. By making the individual (time-series-)properties either a coordinate-, attribute-, or data-variable we have a reasonable control of the behavior under operations. I do not know if this is sufficient in all cases (probably not), but it may be for most.

Reply:
@soininen Also thanks for the merging example. Have a look at https://github.com/mantidproject/dataset/blob/design/test/Workspace2D_test.cpp#L13, this shows a working example of merging (creating an extra dimension), including a temperature coordinate. This sounds very similar to what you describe:
- `MergeRuns` flattens the dimension, whereas with `Dataset::concatenate` it is kept --- and the `TimeSeries` of temperatures would simply be an axis for the new dimension.
- Rejecting of different wavelengths/chopper speeds and summing of monitor and detector counts would probably need to be done as a special implementation for merging the metadata. If it is a separate `Dataset` it is easy: Store wavelength and chopper speeds a coordinate variables, monitor and detector counts as data variables. Then simply attempt adding, this will throw if the coordinates mismatch.